CONTRIBUTING.md

```diff
@@ -0,0 +1,142 @@
+
+In general, error messages passed to exceptions should be a fairly terse description of what happened, and not contain extra diagnostic, contextual, or prescriptive correction advice (see `obj` and `help_text` below for more on that).
+
+## When and how to use AnsibleError
+
+The `AnsibleError` exception type provides support for improved error reporting. However, if no arguments other than a message are given, there's usually no benefit over using built-in exception types. So what are those other arguments?
+
+* `obj` \- Usually a variable responsible for the error being raised – not an `Exception` instance. If this value is `Origin` tagged, then the error message shown to the user will be able to provide context showing what content triggered the error.
+* `help_text` \- Instructions and additional detail that helps the user understand how to resolve the error. By putting that information here, it allows the `message` to be shorter and focus on the problem. This information will be shown *after* the contextual error details provided by `obj`, if any.
+
+## Display warnings and errors
+
+The existing `warning` and `deprecated` methods on the `Display` object now support passing optional `help_text` and `obj` arguments \- matching those on `AnsibleError` for prescriptive guidance and source context from `Origin`\-tagged values. Additionally, the new `error_as_warning` method accepts an exception object and an optional contextual message directly, allowing for a caught exception to be converted to a warning automatically while still preserving the exception detail, traceback, and source object context (where applicable).
+
+## Jinja plugin errors
+
+In Jinja plugins, the `AnsibleFilterError` and `AnsibleLookupError` exception types are no longer needed. Instead, use whatever exception type is appropriate for the error condition. For example, if you detect an incompatible type, raise a `TypeError`.
+
+# Public surface area
+
+Starting with ansible-core 2.19 we're trying to be more intentional about what features are part of our public API and other public surface areas (e.g. CLI, configuration options, module arguments, ansible-core provided Jinja globals, etc.).
+
+## Imports in public Python modules
+
+To make public Python modules more self-documenting:
+
+* Imports in any file considered public API must be sunder-prefixed (e.g. `_module_name`) to avoid confusion about imported objects being part of the public module API.
+* Prefer module-level imports (e.g. `from ansible._internal import _amodule`) with dotted usage (e.g. `foo: _amodule.thing`). This solves many circular import issues and reduces the need for sunder-prefix aliasing on internal imports in public API. Hot code paths can use locals or aliased objects, but sparingly and only where it really matters.
```

Comment:
```suggestion
To make public Python modules more self-documenting:

* Imports in any file considered public API must be sunder-prefixed (for example, `_module_name`) to avoid confusion about imported objects being part of the public module API.
* Prefer module-level imports (for example, `from ansible._internal import _amodule`) with dotted usage (for example, `foo: _amodule.thing`). This solves many circular import issues and reduces the need for sunder-prefix aliasing on internal imports in the public API. Hot code paths can use locals or aliased objects, but sparingly and only where it really matters.
```